### PR TITLE
RED-144 updates for submit exam + response

### PR DESCRIPTION
**Pull request by doombot-exe**
preliminary pull request to compliment branch of the same name on the frontend:
- students will be asked to take a new version if they don't have one open
- students problem states save and can be submitted
- grade results are shown in a modal according to the exam settings
- after an exam is over, problems will be hidden and a new version will need to be started
- backend will not comply when the delay hasn't passed yet, but the front end isn't ready for that
- exceeding the number of allowed graded attempts per version makes the back end err, but the message doesn't make it beyond STDERR
- there are only backend checks against exceeding the allowed number of versions
- an exam version cannot be ended early yet

### CODE REVIEW

File: src/features/courses/course-controller.ts [hunk 1]
```diff
@@ -1150,6 +1165,7 @@ class CourseController {
                 formData = gradeInstance.currentProblemState;
                 sourceFilePath = gradeInstance.webworkQuestionPath;
                 problemSeed = gradeInstance.randomSeed;
+                calculatedRendererParameters.outputformat = OutputFormat.ASSESS;
```
Comment (red-tt-bear) [inline src/features/courses/course-controller.ts:1168]: Why not add this to the getCalculatedRendererParams?
Maybe add an option for isAssessment that returns ASSESS

Also this look like too generic of a way of putting this here, if we ever have a non null workbook (like viewing previous submissions, I'm guessing that this is in one of the get grades and I would assume this would break the workbook veiwer (but I could be wrong)

Either way moving the logic into the calculate renderer params should make it more clear

File: src/features/courses/course-controller.ts [hunk 2]
```diff
@@ -2357,19 +2363,33 @@ class CourseController {
...
 
         try {
             const include: sequelize.IncludeOptions[] = [];
+            const subInclude: sequelize.IncludeOptions[] = [];
+
+            if (!_.isNil(studentTopicAssessmentInfoId)) {
+                subInclude.push({
+                    model: StudentGradeInstance,
+                    as: 'gradeInstances',
+                    required: true,
```
Comment (red-tt-bear) [inline src/features/courses/course-controller.ts:2378]: Why is this required true?
If we missing records does it make sense to return them and then log errors with null checks

File: src/features/courses/course-controller.ts [hunk 3]
```diff
@@ -2399,16 +2419,18 @@ class CourseController {
             const questions = await CourseWWTopicQuestion.findAll(findOptions);
             if (!_.isNil(courseTopicContentId)){
                 const topic = await this.getTopicById({id: courseTopicContentId});
-                // TODO remove assessment hardcoding
+                // TODO remove assessment hardcoding -- userId nil-check is for TS
```
Comment (red-tt-bear) [inline src/features/courses/course-controller.ts:2422]: > nil-check is for TS?
What do you mean by this? In other cases it is because of how TS manages scoping, if you nil check and then put it in a closure TS will be wrong and think it is nillable, but in this case it looks like legit concern

File: src/features/courses/course-controller.ts [hunk 4]
```diff
@@ -2399,16 +2419,18 @@ class CourseController {
...
                 if (topic.topicTypeId === 2 && !_.isNil(userId)) {
                     await questions.asyncForEach(async (question) => {
                         if (_.isNil(question.grades)) throw new RederlyExtendedError('Impossible! Found an assessment question without a grade.');
                         const version = await courseRepository.getCurrentInstanceForQuestion({questionId: question.id, userId});
                         question.webworkQuestionPath = version?.webworkQuestionPath ?? question.webworkQuestionPath;
                         question.grades.randomSeed = version?.randomSeed ?? question.grades.randomSeed;
+                        question.problemNumber = version?.problemNumber ?? question.problemNumber;
```
Comment (red-tt-bear) [inline src/features/courses/course-controller.ts:2429]: Should we do this through the assign with the utility function like we have been doing elsewhere?

**Review by red-tt-bear — CHANGES_REQUESTED** [review]
Please move the Assess renderer param to the calculate
Please move required true to false and log nils unless it is intentional

File: src/features/courses/course-route.ts [hunk 5]
```diff
@@ -451,6 +453,32 @@ router.put('/question/grade/:id',
         }));
     }));
 
+router.put('/question/grade/instance/:id',
```
[the remark this answers is not in the document]
Comment (doombot-exe) [reply src/features/courses/course-route.ts:456]: current problem state

File: src/features/courses/course-controller.ts [hunk 6]
```diff
@@ -2730,7 +2753,7 @@ class CourseController {
                 webworkQuestionPath,
                 randomSeed,
                 userId,
-                problemNumber: problemOrder[index],
+                problemNumber: problemOrder[index]+1, // problemOrder starts from 0
```
Comment (PanopticaRising) [inline src/features/courses/course-controller.ts:2756]: We should consider starting to use explicit functions for when we need to increment the problemNumber, so it's clear when we're using the `raw problemNumber` and a `user-facing problemNumber`

File: src/features/courses/course-route.ts [hunk 7]
```diff
@@ -224,22 +225,22 @@ router.get('/questions',
                 }
 
                 // get version info - descending by startTime unless specific id is included in query
-                let versions: StudentTopicAssessmentInfo[] | undefined;
                 if (_.isNil(req.query.studentTopicAssessmentInfoId)) {
-                    versions = _.orderBy(topic.topicAssessmentInfo.studentTopicAssessmentInfo, ['startTime'],'desc');
+                    // _.orderBy puts in ascending order
+                    const versions = _.orderBy(topic.topicAssessmentInfo.studentTopicAssessmentInfo, ['startTime']).reverse();
```
Comment (PanopticaRising) [inline src/features/courses/course-route.ts:230]: https://lodash.com/docs/4.17.15#orderBy `['
```suggestion
                    const versions = _.orderBy(topic.topicAssessmentInfo.studentTopicAssessmentInfo, ['startTime'], ['desc']);
```

File: src/features/courses/course-controller.ts (hunk 2 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2378]: if the request includes a specific student_topic_assessment_info_id, then there *must* be associated student grade instances. That ID identifies a specific exam version.

File: src/features/courses/course-route.ts (hunk 7 above)
Comment (doombot-exe) [reply src/features/courses/course-route.ts:230]: yeah, I think I was confusing _.sortBy and _.orderBy

File: src/features/courses/course-controller.ts (hunk 6 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2756]: Yes, this is why I used problemOrder to stash the (potentially randomized) sequence - I just didn't see the point of mapping x => x+1 instead of doing it here.

File: src/features/courses/course-controller.ts (hunk 1 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:1168]: this should be moved - agreed - but adding reliance on topicType to the existing logic was a complication where the concern was breaking existing functionality... will handle this in cleanup.

File: src/features/courses/course-controller.ts (hunk 3 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2422]: yeah, I shouldn't have to check it again, but TS was complaining...

File: src/features/courses/course-controller.ts (hunk 4 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2429]: agreed, but this question object is a sequelize with includes -- so it's not overwriting one of our "fundamental" objects.

Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:2430]: Yeah, but that is why we do the multiple interface approach and do the assign using _.pick

File: src/features/courses/course-controller.ts (hunk 2 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:2379]: So you are saying that:
since student topic assessment exists then grade instances must exist

I can agree with that, however, we probably need the nil check anyway and if the generation has a mistake then we will catch it in those nil checks

With required true it would return that student the student grade instance didn't exist and then we would have a red herring of a bug

File: src/features/courses/course-controller.ts (hunk 3 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:2423]: Yes you should, this is not in the closure for the userId check
if userId is nil the above if block will not run, since it is no longer in that if statement it could once again be nil

This is a valid TS complaint

File: src/features/courses/course-controller.ts (hunk 1 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:1169]: I'm fine with that

However my concern is more so with adding extra functionality that will change the existing

We've been kicking a lot of stuff downstream and the result is that either:
1. We will forget about it
2. Higher priority tasks will build up between where we will build this into a deeper level of code that is harder to touch
3. Bugs / problems will come of it before we get there (i.e. if we do this outside of the calculate and need it in multiple places but do it differently)

File: src/features/courses/course-controller.ts (hunk 2 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2379]: I am saying that the student-topic-assessment-info and its corresponding student-grade-instance children are created in the same transaction. You cannot have either independently.

So, the recommendation is a nil check on the result of this query before returning?

File: src/features/courses/course-controller.ts (hunk 4 above)
Comment (doombot-exe) [reply src/features/courses/course-controller.ts:2430]: Oh right - those are already implemented, I just forgot to take advantage!

File: src/features/courses/course-controller.ts (hunk 2 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:2379]: not before returning, the type has it as an optional field so when you go to access it you would have to nil check there anyway, I'm saying we have one of our classic impossible errors be thrown at that point

Although I agree that currently, it is not possible with the transaction all it takes is either:
* for someone to delete one of the awaits by accident and have it as an orphaned promise thus not respecting the transaction (starting a new one)
* for some change to come in that someone creates these objects independently and do it incorrect

**Review by red-tt-bear — APPROVED** [review]

File: src/features/courses/course-route.ts [hunk 8]
```diff
@@ -251,7 +254,8 @@ router.get('/questions',
 
         const questions = await courseController.getQuestions({
             userId: userId,
-            courseTopicContentId: req.query.courseTopicContentId
+            courseTopicContentId: req.query.courseTopicContentId,
+            studentTopicAssessmentInfoId: req.query.studentTopicAssessmentInfoId ?? version?.id,
```
Comment (PanopticaRising) [inline src/features/courses/course-route.ts:258]: This will cause an error to get thrown if `version` is null, right? So this should be fine?
